static/app/components/notificationActions/notificationActionManager.tsx

```diff
@@ -0,0 +1,203 @@
+   */
+  updateAlertCount: (projectId: number, alertCount: number) => void;
+  /**
+   * Optional list of roles to display as recipients of Sentry notifications
+   */
+  recipientRoles?: string[];
+};
+
+const NotificationActionManager = ({
+  actions,
+  availableActions,
+  recipientRoles,
+  project,
+  updateAlertCount,
+}: NotificationActionManagerProps) => {
+  const [notificationActions, setNotificationActions] =
+    useState<Partial<NotificationAction>[]>(actions);
+
+  const addNotificationAction = (
+    actionInstance: AvailableNotificationAction['action']
+  ) => {
+    const updatedActions = [...notificationActions, actionInstance];
+    setNotificationActions(updatedActions);
+    updateAlertCount(parseInt(project.id, 10), updatedActions.length);
+  };
+
+  const removeNotificationAction = (index: number) => {
+    // Removes notif action from state using the index
+    const updatedActions = [...notificationActions];
+    updatedActions.splice(index, 1);
+    setNotificationActions(updatedActions);
+    updateAlertCount(parseInt(project.id, 10), updatedActions.length);
+  };
+
+  const updateNotificationAction = (index: number, updatedAction: NotificationAction) => {
+    // Updates notif action from state using the index
+    const updatedActions = [...notificationActions];
+    updatedActions.splice(index, 1, updatedAction);
+    setNotificationActions(updatedActions);
+  };
+
+  // Lists the available actions for each service
+  const availableServices: Record<
+    NotificationActionService,
+    AvailableNotificationAction[]
+  > = {
+    [NotificationActionService.SENTRY_NOTIFICATION]: [],
+    [NotificationActionService.EMAIL]: [],
+    [NotificationActionService.SLACK]: [],
+    [NotificationActionService.PAGERDUTY]: [],
+    [NotificationActionService.MSTEAMS]: [],
+    [NotificationActionService.SENTRY_APP]: [],
+  };
+  availableActions.forEach(a => {
+    availableServices[a.action.serviceType as NotificationActionService].push(a);
+  });
+
+  // Groups the notification actions together by service
+  // Will render the notif actions in the order the keys are listed in
+  const actionsMap: Record<
+    NotificationActionService,
+    {action: NotificationAction; index: number}[]
+  > = {
+    [NotificationActionService.SENTRY_NOTIFICATION]: [],
+    [NotificationActionService.EMAIL]: [],
+    [NotificationActionService.SLACK]: [],
+    [NotificationActionService.PAGERDUTY]: [],
+    [NotificationActionService.MSTEAMS]: [],
+    [NotificationActionService.SENTRY_APP]: [],
+  };
+  notificationActions.forEach((action, index) => {
+    if (action.serviceType) {
+      actionsMap[action.serviceType].push({action, index});
+    }
+  });
+
+  // Groups the pagerduty integrations with their corresponding allowed services
+  const pagerdutyIntegrations: Record<number, AvailableNotificationAction[]> = {};
+  availableServices[NotificationActionService.PAGERDUTY].forEach(service => {
```

Comment:
anytime we're looping over to populate a constante, lets stick in a `useMemo` at least

Reply:
this will never change from the initial render unless availableServices changes, and availableServices can only change if we refetch the data for availableActions (which doesn't happen unless you reload the page)

Reply:
in general this component doesn't rerender apart from the initial render

Reply:
👍 awesome, then in that case lets memoize it.
It's a cheap optimization so 🤷